File: baekjoon/basic2/2468/2468.cpp
// 백준 2468 : 안전 영역 (https://www.acmicpc.net/problem/2468)

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
int n, ny, nx;
const int N = 100;

int visited[N][N];
int a[N][N];
int mp[N][N];

int dy[4] = {0,0,1,-1};
int dx[4] = {1,-1,0,0};

void go(int y, int x){
    visited[y][x] = 1;
    for(int i=0; i<4; i++){
        ny = y + dy[i];
        nx = x + dx[i];
        if (ny >= n || ny < 0 || nx >= n || nx < 0) continue;
        if (a[ny][nx] == 1 && visited[ny][nx] == 0) go(ny, nx);
    }
    return;
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    cin >> n;
    int h = 0;

    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            cin >> mp[i][j];
            h = max(h, mp[i][j]);
        }
    }
    int ret = 0;
    for(int i=0; i<h; i++){
        int cnt = 0;
        for(int j=0; j<n; j++){
            for(int k=0; k<n; k++){
                visited[j][k] = 0;
                if (mp[j][k] <= i) a[j][k] = 0;
                else a[j][k] = 1;    
            }
        }
        for(int j=0; j<n; j++){
            for(int k=0; k<n; k++){
                if (a[j][k] == 1 && visited[j][k] == 0){
                    go(j, k);
                    cnt++;
                }
            }
        }
        ret = max(ret, cnt);
    }

    cout << ret << "\n";
    
    return 0;
} 

/*
#include <iostream>
#include <algorithm>
using namespace std;
const int mn = 101;
int a[mn][mn], visited[mn][mn];
int n, ret;
int ny, nx, cnt;
int dy[4] = {0,0,1,-1};
int dx[4] = {1,-1,0,0};


void dfs(int y, int x, int d){
    visited[y][x] = 1;
    for(int i=0; i<4; i++){
        ny = y + dy[i];
        nx = x + dx[i];
        if(ny>=n || n<0 || nx>=n || nx<0) continue;
        if(!visited[ny][nx] && a[ny][nx] > d){
            dfs(ny, nx, d);
        }
    }
    return;
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    cin >> n;
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            cin >> a[i][j];
        }
    }
    for(int d=0; d<101; d++){
        fill(&visited[0][0], &visited[0][0] + mn*mn, 0);
        cnt = 0;
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(!visited[i][j] && a[i][j] > d){
                    dfs(i, j, d);
                    cnt++;
                }
            }
        }
        ret = max(ret, cnt);
    }
    cout << ret << "\n";
}
*/

File: baekjoon/basic2/2468/2468.py
# 백준 2468 안전 영역 (https://www.acmicpc.net/problem/2468)
import sys
sys.setrecursionlimit(10**5)

dy = [0,0,1,-1]
dx = [1,-1,0,0]

def dfs(y, x):
    visited[y][x] = 1
    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i] 
        if ny >= n or ny < 0 or nx >= n or nx < 0:
            continue
        if visited[ny][nx] == 0 and mp[ny][nx] == 1:
            dfs(ny, nx)
    return

def dfs_stack(y, x):
    visited[y][x] = 1
    stack = []
    stack.append((y, x))

    while stack:
        cy, cx = stack.pop()
        for i in range(4):
            ny = cy + dy[i]
            nx = cx + dx[i] 
            if ny >= n or ny < 0 or nx >= n or nx < 0:
                continue
            if visited[ny][nx] == 0 and mp[ny][nx] == 1:
                stack.append((ny, nx))
                visited[ny][nx] = 1
    return

n = int(sys.stdin.readline())
a = []
h = 0
for i in range(n):
    s = list(map(int, sys.stdin.readline().split()))
    h = max(h, max(s))
    a.append(s)

mp = [[0] * n for _ in range(n)]

ret = 0
for rain in range(h):
    visited = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if a[i][j] <= rain:
                mp[i][j] = 0
            else:
                mp[i][j] = 1
    cnt = 0
    for i in range(n):
        for j in range(n):
            if visited[i][j] == 0 and mp[i][j] == 1:
                dfs_stack(i, j)
                cnt += 1
    ret = max(ret, cnt)
sys.stdout.write(str(ret)+"\n")


